swap/chain/persistentqueue.go

```diff
@@ -0,0 +1,121 @@
+package chain
+
+import (
+	"context"
+	"encoding"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/swarm/state"
+)
+
+/*
+	PersistentQueue represents a queue stored in a state store
+  Items are enqueued by writing them to the state store with the timestamp as prefix and a nonce in case two items get queued at the same time
+  It provides a (blocking) Next function to wait for a new item to be available. Only a single call to Next may be active at any time
+	To allow atomic operations with other state store operations all functions only write to batches instead of writing to the store directly
+	The user must ensure that all functions (except Next) are called with the same lock held which is provided externally so multiple queues can use the same
+	The queue provides no dequeue function. Instead an item must be deleted by its key
+*/
+
+// PersistentQueue represents a queue stored in a state store
+type PersistentQueue struct {
+	store   state.Store   // the store backing this queue
+	prefix  string        // the prefix for the keys for this queue
+	trigger chan struct{} // channel to notify the queue that a new item is available
+	nonce   uint64        // increasing nonce. starts with 0 on every startup
+}
+
+// NewPersistentQueue creates a structure to interact with a queue with the given prefix
+func NewPersistentQueue(store state.Store, prefix string) *PersistentQueue {
+	return &PersistentQueue{
+		store:   store,
+		prefix:  prefix,
+		trigger: make(chan struct{}, 1),
+		nonce:   0,
+	}
+}
+
+// Queue puts the necessary database operations for enqueueing a new item into the supplied batch
+// It returns the generated key and a trigger function which must be called once the batch was successfully written
+// This only returns an error if the encoding fails which is an unrecoverable error
+// A lock must be held and kept until after the trigger function was called or the batch write failed
+func (pq *PersistentQueue) Queue(b *state.StoreBatch, v interface{}) (key string, trigger func(), err error) {
```

Comment:
Then it is a programming error and the behaviour is undefined. Most likely it wouldn't affect any calls to `Next` which happen afterwards, but a `Next` which is already blocking would not wake up.

Reply:
Could you add this as a comment?
e.g.:
```golang
// If trigger is called before writing the batch, it may cause an already blocking call to Next to not wake up or other undefined behaviours
```
On top of this, perhaps also add such a comment when you are actually calling this function (for example in `finalizeRequestCancelled-txqueue.go:419`: 
`// call trigger function after writing to the batch to prevent undefined behaviour`)